File: scripts/cfg_server.py
#!/usr/bin/env python3
# encoding: utf-8

import sys
from typing import Final

import rospy
from dynamic_reconfigure.server import Server

# custom config
from catkin_template.cfg import ConfigConfig

# constants
NODE_NAME: Final[str] = 'py_cfg_server'


def callback(config: ConfigConfig, _) -> ConfigConfig:
  rospy.loginfo("Reconfigure: {int_param}, {double_param}, {str_param}, {bool_param}, {size}".format(**config))

  # config validation ... 

  return config


def main() -> None:

  # initialize ROS Node
  rospy.init_node(NODE_NAME, sys.argv)

  server = Server(ConfigConfig, callback)
  
  rospy.spin()


if __name__ == '__main__':
    main()

File: src/cfg/cfg_server.cpp
#include <cstdint> // uint32_t

#include <boost/bind.hpp>

#include <ros/ros.h> // NodeHandle, Publisher, Rate
#include <dynamic_reconfigure/server.h>

// custom config
#include <catkin_template/ConfigConfig.h>

// constants
static constexpr auto kNodeName = "cpp_cfg_server";


void callback(catkin_template::ConfigConfig& config, std::uint32_t level) {
  ROS_INFO("Reconfigure: %d, %.1f, %s, %s, %d, level = %d",
           config.int_param, config.double_param, config.str_param.data(),
           config.bool_param? "True":"False", config.size, level);

  // validation ...
}

int main(int argc, char **argv) {

  // initialize ROS Node
  ros::init(argc, argv, kNodeName);

  auto server = dynamic_reconfigure::Server<catkin_template::ConfigConfig>();
  auto callback_type = dynamic_reconfigure::Server<catkin_template::ConfigConfig>::CallbackType();

  server.setCallback(boost::bind(&callback, boost::placeholders::_1, boost::placeholders::_2));

  ros::spin();

  return 0;
}